lib/bolt/plugin/pkcs7.rb

```diff
@@ -0,0 +1,76 @@
+# frozen_string_literal: true
+
+require 'bolt/secret/base'
+require 'fileutils'
+
+module Bolt
+  class Plugin
+    class Pkcs7 < Bolt::Secret::Base
+      def name
+        'pkcs7'
+      end
+
+      def initialize(boltdir, options)
+        require 'openssl'
+        @boltdir = boltdir
+        @options = options || {}
+      end
+
+      def private_key_path
+        path = @options['private-key'] || 'keys/private_key.pkcs7.pem'
+        File.absolute_path(path, @boltdir)
+      end
+
+      def private_key
+        @private_key ||= OpenSSL::PKey::RSA.new(File.read(private_key_path))
+      end
+
+      def public_key_path
+        path = @options['public-key'] || 'keys/public_key.pkcs7.pem'
+        File.absolute_path(path, @boltdir)
+      end
+
+      def public_key
+        @public_key ||= OpenSSL::X509::Certificate.new(File.read(public_key_path))
+      end
+
+      def keysize
+        @options['keysize'] || 2048
+      end
+
+      # The following implementations are intended to be compatible with hiera-eyaml
+      def encrypt_value(plaintext)
+        cipher = OpenSSL::Cipher::AES.new(256, :CBC)
+        OpenSSL::PKCS7.encrypt([public_key], plaintext, cipher, OpenSSL::PKCS7::BINARY).to_der
+      end
+
+      def decrypt_value(ciphertext)
+        pkcs7 = OpenSSL::PKCS7.new(ciphertext)
+        pkcs7.decrypt(private_key, public_key)
+      end
+
+      def secret_createkeys
+        key = OpenSSL::PKey::RSA.new(keysize)
+
+        cert = OpenSSL::X509::Certificate.new
+        cert.subject = OpenSSL::X509::Name.parse('/')
+        cert.serial = 1
+        cert.version = 2
+        cert.not_before = Time.now
+        cert.not_after = Time.now + 50 * 365 * 24 * 60 * 60
+        cert.public_key = key.public_key
+        cert.sign(key, OpenSSL::Digest.new('SHA512'))
+
+        private_keydir = File.dirname(private_key_path)
+        FileUtils.mkdir_p(private_keydir) unless File.exist?(private_keydir)
+        FileUtils.touch(private_key_path)
+        FileUtils.chmod(0o600, private_key_path)
```

Comment:
This appears to not work on windows. What should we do?

Reply:
https://github.com/puppetlabs/bolt/blob/ea5b61a8daf363d679ed21225564dfb4fab6ad79/lib/bolt/transport/local_windows.rb#L59 maybe?